mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h

```diff
@@ -1824,12 +1870,40 @@ template <typename T> struct Conv2DRewritePattern : public OpRewritePattern<T> {
         padGemmInputTransform.passThrough("gemmN");
       }
 
-      TransformMapAttr padGemmInputTransformAttr = padGemmInputTransform.get();
+      padGemmInputTransformAttr = padGemmInputTransform.get();
       gemmInputPad = b.create<miopen::TransformOp>(loc, gemmInput,
                                                    padGemmInputTransformAttr);
       // input padding end
     }
 
+    // KPack for input tensor.
+    Value gemmInputKPack = gemmInputPad;
+
+    // FIXME. consider backward convolution.
+    if ((KPack > 1) && (convOpType == miopen::ConvOpType::Fwd)) {
+      BottomUpCTBuilder sourceTransform =
+          (isInputPad) ? padGemmInputTransform : gemmInputTransform;
+      TransformMapAttr sourceTransformAttr =
+          (isInputPad) ? padGemmInputTransformAttr : gemmInputTransformAttr;
+      Value source = (isInputPad) ? gemmInputPad : gemmInput;
+
+      BottomUpCTBuilder kpackGemmInputTransform =
+          BottomUpCTBuilder::above(sourceTransform, sourceTransformAttr);
+      // Passthrough gemmG (dim 0) and gemmN (dim 2).
+      kpackGemmInputTransform.passThrough(sourceTransform.endName(0));
+      kpackGemmInputTransform.passThrough(sourceTransform.endName(2));
+      // Use Unmerge to split gemmK (dim 1) into gemmK and gemmKPack, place
+      // gemmKPack at dim 3.
+      int64_t gemmKLength = sourceTransform.endSize(sourceTransform.endName(1));
```

Comment:
Did I not add an `endSize()` that takes an index` making this `sourceTransform.endSize(1);`? If not ... whoops.

Reply:
There is indeed an `endSize(uint32_t dim)` interface. I was just use all names here. Will fix that in a separate cleanup PR.